tools/gen-intrinsics-specs.py

```diff
@@ -0,0 +1,718 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2021, Arm Limited
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+from tabulate import tabulate
+import argparse
+import csv
+import doctest
+
+
+def quote_literal(val):
+    if len(val) > 0:
+        return f"``{val}``"
+    return val
+
+
+rst_header_levels = ['=', '~', '-', '_', '@']
+__CSV_SECTION_PREFIX = '<SECTION>'
+__CSV_COMMENT_PREFIX = '<COMMENT>'
+__CSV_HEADER_PREFIX = '<HEADER>'
+__intrinsic_table_header = ['Intrinsic', 'Argument preparation',
+                            'AArch64 Instruction', 'Result', 'Supported architectures']
+
+__INTRINSIC_TABLE_KEYWORD = '__intrinsic_table'
+__SECTION_TEXT_KEYWORD = '__section_text'
+
+__SECTION_KEYWORDS = [__INTRINSIC_TABLE_KEYWORD, __SECTION_TEXT_KEYWORD]
+
+
+def rst_literal_quote(mapping):
+    r"""
+    >>> rst_literal_quote('a; b')
+    ' ::\n\n    a \n    b \n\n'
+
+    >>> rst_literal_quote('a b')
+    ' ::\n\n    a b \n\n'
+
+    >>> rst_literal_quote('')
+    ''
+    """
+    if mapping == "":
+        return ""
+
+    lines = mapping.split(';')
+    indented_lines = [f"    {line.strip()} " for line in lines]
+    lines = [" ::", ""] + indented_lines + ["\n"]
+    return '\n'.join(lines)
+
+
+def quote_split_intrinsics(intrinsic):
+    r"""
+    >>> quote_split_intrinsics('int f(int x, float y)')
+    '.. code:: c\n\n    int f(\n        int x,\n        float y)'
+
+    >>> quote_split_intrinsics('int f(int x)')
+    '.. code:: c\n\n    int f(int x)\n'
+    """
+    intrinsic_without_ending = intrinsic.removesuffix(')')
```

Comment:
The `str.removesuffix` method doesn't exist before Python 3.9, which means this will fail on a lot of currently deployed Linux distros. Even the latest Ubuntu LTS (20.04) is still on 3.8.5.

I suggest either making this script compatible with older versions of Python, or if not, checking whether `sys.version_info >= (3,9)` at startup and generating a more friendly error message than `AttributeError: 'str' object has no attribute 'removesuffix'`.

Reply:
I have chosen a different syntax, which should be compatible with many more versions of python.